developer-docs/providers/implementers-guide.md

```diff
@@ -616,7 +616,147 @@ None.
+   passed as `news`. Note the possible confusion: `olds` and `oldInputs` represent the candidate state in this call,
+   not the state prior to the `pulumi refresh` operation.
+
+5. If the provider responds with `DIFF_NONE` to the `Diff` call, Pulumi assumes that no drift has occurred and any
+   differences, if any, between the candidate state and the program are immaterial. Pulumi omits the resource from diff
+   display. The user receives a message that nothing has changed. However, while resource outputs are indeed unchanged,
+   resource inputs are silently rewritten in the state to the value of inputs returned by `Read`, the candidate inputs.
+
+   This behavior is similar to `pulumi up` that will rewrite the inputs in state to the newly checked inputs even if
+   the provider responds with `DIFF_NONE` to the `Diff` call.
+
+6. If the provider responds with `DIFF_SOME` to the `Diff` call, Pulumi renders the changes to the user asking for a
+   confirmation. Once the user confirms the operation, the candidate state is written as the actual state to the state
+   store.
+
+For a concrete example, suppose we have an AWS S3 bucket with tag "a" that has drifted to have tag "b" in the cloud.
+
+Pulumi will call the provider `Read` method, passing "a" as old state. In the case of S3 buckets, tags are modeled as
+both inputs and (matching) outputs, duplicating the information. Slightly simplifying, the call will look like this:
+
+```shell
+{
+  "id": "my-bucket-ae91e13",
+  "name": "my-bucket",
+  "type": "aws:s3/bucketV2:BucketV2",
+  "urn": "urn:pulumi:dev::document-refresh::aws:s3/bucketV2:BucketV2::my-bucket",
+  "properties": {
+    "tags": {
+      "tagName": "a"
+    }
+  },
+  "inputs": {
+    "tags": {
+      "tagName": "a"
+    }
+  }
+}
+```
+
+
+The provider will locate the bucket in the cloud, and determine that the real tag value is "b". It will respond to
+`Read` call with the output (properties) and input values as found in the cloud:
+
+```json
+{
+  "id": "my-bucket-ae91e13",
+  "properties": {
+    "tags": {
+      "tagName": "b"
+    }
+  },
+  "inputs": {
+    "tags": {
+      "tagName": "b"
+    }
+  }
+}
+```
+
+The results of the `Read` method constitute the new candidate state. The engine will then ask the provider to compare
+the candidate state against the program inputs. This gives the provider a chance to respond with a DIFF_NONE indicating
+that the changes discovered by `Read` are not essential. It also allows the provider to do some domain-specific
+formatting on the diff to improve how it is displayed in Pulumi CLI.
+
+Following our example, `Diff` call may look like this:
+
+```json
+{
+  "id": "my-bucket-ae91e13",
+  "name": "my-bucket",
+  "type": "aws:s3/bucketV2:BucketV2",
+  "urn": "urn:pulumi:dev::document-refresh::aws:s3/bucketV2:BucketV2::my-bucket",
+  "olds": {
+    "tags": {
+      "tagName": "b"
+    }
+  },
+  "oldInputs": {
+    "tags": {
+      "tagName": "b"
+    }
+  },
+  "news": {
```

Comment:
It's not from the program, which is not being read, it's the inputs from the state before refresh. Must be.. I'll amend.